thirdparty/DiskANN/src/pq_flash_index.cpp

```diff
@@ -332,32 +338,47 @@ namespace diskann {
       return;
     }
 
-    std::vector<int64_t> tmp_result_ids_64(sample_num, 0);
-    std::vector<float>    tmp_result_dists(sample_num, 0);
+    int64_t tmp_result_ids_64;
+    float tmp_result_dists;
 
-#pragma omp parallel for schedule(dynamic, 1) num_threads(nthreads)
-    for (_s64 i = 0; i < (int64_t) sample_num; i++) {
-      cached_beam_search(samples + (i * sample_aligned_dim), 1, l_search,
-                         tmp_result_ids_64.data() + (i * 1),
-                         tmp_result_dists.data() + (i * 1), beamwidth);
+    auto id = 0;
+    while (this->search_counter.load() < sample_num && id < sample_num &&
+          !this->semaph.IsWaitting()) {
+      cached_beam_search(samples + (id * sample_aligned_dim), 1, l_search,
+                        &tmp_result_ids_64, &tmp_result_dists, beamwidth);
+      id++;
     }
 
+    if (this->semaph.IsWaitting()) {
+      this->semaph.Signal();
+      return;
+    }
+
+    this->count_visited_nodes.exchange(false);
```

Comment:
Simply moving this line doesn't help guarantee "stop updating `this->search_counter` and `this->node_visit_counter` once we have reached `sample_num` searches".

But since the stop condition `this->search_counter.load() == sample_num` is not strict(it's ok to search more than `sample_num` times to building the cache), we can keep the code as is. But please add a comment like "it is intentional and ok that search_counter/node_visit_counter may continue being updated after search_counter has reached sample_num before we flip count_visited_nodes".

Reply:
if count_visited_nodes == true： diskann will update node_visit_counter in each iteration, and update search_counter in each query. So when generate cache task done,  i want to release node_visit_counter memory and reduce some search time from updating node_visit_counter and search_counter.

Reply:
My point is:

In general sync mechanisms are used to guarantee strong correctness, but here we use locks/atomics while tolerating a degree of inaccuracy(as you mentioned in another comment that the update of the counter is not strict). So we should leave a comment explicitly stating such.